src/dmd/semantic3.d

```diff
@@ -1411,6 +1413,47 @@ private extern(C++) final class Semantic3Visitor : Visitor
                 }
             }
 
+            // Reset the instance to T.init to prevent double destruction for
+            // dynamically allocated instances, e.g. when new T() throws
+            {
+                auto this_ = new ThisExp(ctor.loc);
+                auto other = new DsymbolExp(ctor.loc, ad);
+                Expression reset;
+
+                // Rewrite as (cast(void*) this)[0 .. size] = typeid(this).initializer[]
```

Comment:
The codegen layer could handle this much more efficiently (no TypeInfo needed, simple blit from init symbol), but as this is only done if the ctor throws, it shouldn't matter.

Reply:
Agreed. I've also thought about a Druntime hook similar to `emplaceInitializer`